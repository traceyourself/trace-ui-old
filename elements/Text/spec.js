import React from 'react'
import renderer from 'react-test-renderer'
import Text from './'

test('Text renders like a boss', () => {
  const component = renderer.create(
    <Text>like a boss</Text>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Text renders transform', () => {
  const component = renderer.create(
    <Text transform='uppercase'>like a boss</Text>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
