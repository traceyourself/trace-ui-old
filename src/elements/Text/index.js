import React, { PropTypes as pt } from 'react'
import { css } from 'glamor'

import { pxToEm } from '../../utils'
import { color } from '../../style'
import { typography } from '../../settings'

const { keys } = Object

css.fontFace(typography.museoSans)

const baseStyle = css({
  lineHeight: 1.5,
  fontFamily: 'Museo Sans'
})

const typeConfig = {
  p: { size: 18, opacity: 0.6 },
  h1: { size: 36, weight: 900 },
  h2: { size: 31.5, weight: 900 },
  h3: { size: 27, weight: 900 },
  h4: { size: 24, weight: 900 },
  h5: { size: 21, weight: 900 },
  h6: { size: 18, weight: 900 },
  label: { size: 16, weight: 900 },
  span: { size: 16 },
  small: { size: 12 }
}

/*
  make: {
    p: css({}),
    h1: css({}),
    ...
   }
 */
const styleConfig = keys(typeConfig)
  .reduce((styles, type) => ({
    ...styles,
    [type]: css({
      fontSize: pxToEm(typeConfig[type].size),
      fontWeight: typeConfig[type].weight || 500,
      opacity: typeConfig[type].opacity || 0.8
    })
  }), {})

const decoration = {
  none: css({textDecoration: 'none'}),
  underline: css({textDecoration: 'underline'}),
  overline: css({textDecoration: 'overline'}),
  'line-through': css({textDecoration: 'line-through'})
}

const transform = {
  none: css({textTransform: 'none'}),
  uppercase: css({textTransform: 'uppercase'}),
  lowercase: css({textTransform: 'lowercase'}),
  capitalize: css({textTransform: 'capitalize'}),
  'full-width': css({textTransform: 'full-width'})
}

const whiteListedProps = {
  id: value => ({id: value}),
  baseRef: value => ({ref: value}),
  decoration: value => decoration[value] || {},
  transform: value => transform[value] || {}
}

const Text = ({
  type: Type = 'span',
  color: clr = 'font',
  children,
  ...rest
}) => {
  const props = keys(whiteListedProps)
    .reduce((props, propName) => {
      if (rest.hasOwnProperty(propName)) {
        return {
          ...props,
          ...whiteListedProps[propName](rest[propName])
        }
      }
      return props
    }, {})

  return (
    <Type {...props} {...baseStyle} {...color[clr]} {...styleConfig[Type]}>
      {children}
    </Type>
  )
}

Text.propTypes = {
  color: pt.oneOf(keys(color)),
  type: pt.oneOf([
    'p',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'label',
    'small'
  ]),
  transform: pt.oneOf(keys(transform)),
  decoration: pt.oneOf(keys(decoration))
}

export default Text
