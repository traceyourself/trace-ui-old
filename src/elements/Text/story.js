import React from 'react'
import { storiesOf } from '@kadira/storybook'
import Text from './index'

storiesOf('Text', module)
  .add('basic', () => (
    <Text>Hello</Text>
  ))
  .add('small', () => (
    <Text type='small'>Hello</Text>
  ))
  .add('paragraph', () => (
    <Text type='p'>
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
    </Text>
  ))
  .add('heading', () => (
    <div>
      <Text type='h1'>
        h1
      </Text>
      <Text type='h2'>
        h2
      </Text>
      <Text type='h3'>
        h3
      </Text>
      <Text type='h3'>
        h3
      </Text>
      <Text type='h4'>
        h4
      </Text>
      <Text type='h5'>
        h5
      </Text>
      <Text type='h6'>
        h6
      </Text>
    </div>
  ))
  .add('colors', () => (
    <div style={{backgroundColor: 'black'}}>
      <Text color='link'>link</Text>
      <Text color='warning'>warning</Text>
      <Text color='success'>success</Text>
      <Text color='danger'>danger</Text>
      <Text color='white'>white</Text>
    </div>
  ))
  .add('transform', () => (
    <div>
      <Text>None: </Text><Text transform='none'>None</Text>
      <br/>
      <Text>Uppercase: </Text><Text transform='uppercase'>Uppercase</Text>
      <br/>
      <Text>Lowercase: </Text><Text transform='lowercase'>Lowercase</Text>
      <br/>
      <Text>capitalize: </Text><Text transform='capitalize'>capitalize</Text>
    </div>
  ))
  .add('decoration', () => (
    <div>
      <Text decoration='underline'>underline</Text>
      <br/>
      <Text decoration='overline'>overline</Text>
    </div>
  ))
