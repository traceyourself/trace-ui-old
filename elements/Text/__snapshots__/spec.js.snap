exports[`test Text renders like a boss 1`] = `
<span
  data-css-1w5ne0l=""
  data-css-bytx6u=""
  data-css-pc0l6w="">
  like a boss
</span>
`;

exports[`test Text renders transform 1`] = `
<span
  data-css-1w5ne0l=""
  data-css-459srw=""
  data-css-bytx6u=""
  data-css-pc0l6w="">
  like a boss
</span>
`;
